CMakeLists.txt

```diff
@@ -1,6 +1,9 @@
 cmake_minimum_required(VERSION 3.5)
 project(hashmap VERSION 1.0.0 LANGUAGES C)
 
+# Option to build the hashmap tests
+option(BUILD_TESTS "Build tests" OFF)
```

Comment:
Actually, in the example of Pablo Arias (I liked the article), he is using the GoogleTest (Specially dedicated for C++ code) to execute the tests. But in our case David Leeds just wrote a custom binary in C to execute some performance tests in his own manner which is grate. So I just simplified the build file to directly build the tests without GoogleTest.
The reason why I added an option for building tests is the additional flag (HASHMAP_METRICS) which is required to build the tests (See the original build file).
To summarize, when the option BUILD_TESTS is ON, we build the library with the flag HASHMAP_METRICS, we build the test binary and we link it with the library (built with the HASHMAP_METRICS flag). And when the BUILD_TESTS is OFF, we just build the library without the HASHMAP_METRICS flag. So the option BUILD_TESTS is necessary.

Reply:
Okay, I understand. If so, we should build two variants of hashmap (two separate targets):
 - `hashmap`
 - `hashmap_metrics`
The first one is just normal build. The second one is hashmap built with `-DHASHMAP_METRICS`.

Also, I didn't say anything about GoogleTest. I mention only three last lines of that file. So, there should be two options:
 - `BUILD_METRICS` "Build library hashmap_metrics"
 - `BUILD_TESTS` "Build tests"

I think, `BUILD_TESTS` should be enabled by default. Also, if `BUILD_TESTS` is enabled, then `BUILD_METRICS` should be forcefully enabled.

This approach should be flexible enough (users can use metrics library if they need metrics).

Reply:
Okay, It a good idea to create two separate targets (hashmap and hashmap_metric) so that we can enable building test by default. But the hashmap_metrics should be just used by the test binary so no need to export it like the "hashmap" (No need for a custom option for it). Actually, the flag HASHMAP_METRICS is just used in conjunction with the test. So normal users (including developer) will not use it (except to make contribution to this project, so build from source).

Reply:
I tried to find a way to clone the hashmap target but I didn't find a clean way to do it. Two solutions are offered to me : 
- Copy the whole hashmap config (source, include, compile options) manually.
- Copy the whole hashmap config dynamically using the following code : 
```cmake
if(BUILD_TESTS)
    # Clone hashmap library properties
    get_target_property(HASHMAP_SOURCES hashmap SOURCES)
    get_target_property(HASHMAP_INCLUDE_DIRECTORIES hashmap INCLUDE_DIRECTORIES)
    get_target_property(HASHMAP_COMPILE_OPTIONS hashmap COMPILE_OPTIONS)
    add_library(hashmap_metrics ${HASHMAP_SOURCES})
    target_include_directories(hashmap_metrics PUBLIC ${HASHMAP_INCLUDE_DIRECTORIES})
    target_compile_options(hashmap_metrics PRIVATE ${HASHMAP_COMPILE_OPTIONS})

    # Add HASHMAP_METRICS flag
    target_compile_definitions(hashmap_metrics PUBLIC HASHMAP_METRICS)
endif()
```

Reply:
So... Then we need to put sources (and compile flags) to variables (`HASHMAP_SOURCES`, `HASHMAP_COMPILE_FLAGS`) and make 2 copies of code (one for `hashmap`, one for `hashmap_metrics`). `BUILD_METRICS` is still needed, I think.

Reply:
There should be ability to build metrics library in any case.

Reply:
So, we should export also hashmap_metrics with an export name like HashMap::HashMapMetrics or HashMap::HashMap_metrics or HashMap::HashMap::Metrics

Reply:
I think, the third one is the best.

Reply:
@Tarik02 It was a pleasure to discuss with you, and I hope you find it as instructive as it was for me. Actually, I will not be able to work on it these days, but I will get back to you as soon as possible.

Reply:
@aminosbh, thank you! It was interesting to me too. You may work on it as soon as you want.